wdl-engine/src/stdlib/read_object.rs

```diff
@@ -0,0 +1,229 @@
+
+    #[test]
+    fn read_object() {
+        let mut env = TestEnv::default();
+        env.write_file("empty.tsv", "");
+        env.write_file("object.tsv", "foo\tbar\nbaz\tqux");
+        env.write_file("too-many-lines.tsv", "foo\tbar\nbaz\tqux\njam\tcakes\n");
+        env.write_file("too-few-lines.tsv", "foo\tbar\n");
+        env.write_file("too-few-columns.tsv", "foo\tbar\nbaz\n");
+        env.write_file("too-many-columns.tsv", "foo\tbar\nbaz\tqux\twrong\n");
+        env.write_file("duplicate.tsv", "foo\tbar\tfoo\nbaz\tqux\tfoo\n");
+        env.write_file("invalid-name.tsv", "foo\tbar-wrong\tfoo\nbaz\tqux\tfoo\n");
+
+        let diagnostic = eval_v1_expr(&mut env, V1::Two, "read_object('empty.tsv')").unwrap_err();
+        assert!(
+            diagnostic.message().contains(
+                "call to function `read_object` failed: expected exactly two lines in file"
+            )
+        );
+
+        let diagnostic =
+            eval_v1_expr(&mut env, V1::Two, "read_object('too-many-lines.tsv')").unwrap_err();
+        assert!(
+            diagnostic.message().contains(
+                "call to function `read_object` failed: expected exactly two lines in file"
+            )
+        );
+
+        let diagnostic =
+            eval_v1_expr(&mut env, V1::Two, "read_object('too-few-lines.tsv')").unwrap_err();
+        assert!(
+            diagnostic.message().contains(
+                "call to function `read_object` failed: expected exactly two lines in file"
+            )
+        );
+
+        let diagnostic =
+            eval_v1_expr(&mut env, V1::Two, "read_object('too-many-columns.tsv')").unwrap_err();
+        assert!(
+            diagnostic
+                .message()
+                .contains("call to function `read_object` failed: line 2 of file")
+        );
+        assert!(
+            diagnostic
+                .message()
+                .contains("does not contain the expected number of columns")
+        );
+
+        let diagnostic =
+            eval_v1_expr(&mut env, V1::Two, "read_object('too-few-columns.tsv')").unwrap_err();
+        assert!(
+            diagnostic
+                .message()
+                .contains("call to function `read_object` failed: line 2 of file")
+        );
+        assert!(
+            diagnostic
+                .message()
+                .contains("does not contain the expected number of columns")
+        );
+
+        let diagnostic =
+            eval_v1_expr(&mut env, V1::Two, "read_object('duplicate.tsv')").unwrap_err();
+        assert!(diagnostic.message().contains(
+            "call to function `read_object` failed: duplicate column name `foo` in line 1 of file"
+        ));
+
+        let diagnostic =
+            eval_v1_expr(&mut env, V1::Two, "read_object('invalid-name.tsv')").unwrap_err();
+        assert!(diagnostic.message().contains(
+            "call to function `read_object` failed: column name `bar-wrong` in line 1 of file"
```

Comment:
Can this error message be better? It appears to be an incorrect column name.

Reply:
it's split this way because it contains the path to the file it failed to read, which in this case are always in the temp directory:

```
call to function `read_object` failed: column name `bar-wrong` in line 1 of file `/var/folders/9t/yj3jt2pn5zx5mzpy6hqc8f540000gp/T/.tmpsDwkHf/invalid-name.tsv` is not a valid WDL object field name
```

Happy to reword it as it is definitely clunky.